A new usage rule for natrec

oskeri opened the pull request:
Also includes some infrastructure changes regarding how the usage rule for natrec is chosen. This decision is now moved to `Usage-restrictions` from `Modality-variant`. 

Consequently, `Modality-variant` is now only used to determine whether mode `𝟘ᵐ` is allowed or not. Perhaps is would make sense to move this to `Usage-restrictions` as well at some point?

nad commented on Graded/Context.agda:
```diff
@@ -125,3 +128,9 @@ _⊛ᶜ_▷_ :
+Greatest-such-thatᶜ : ∀ {ℓ} → (Conₘ n → Set ℓ) → Conₘ n → Set (a ⊔ ℓ)
+Greatest-such-thatᶜ P γ = P γ × (∀ δ → P δ → δ ≤ᶜ γ)
+
+Greatest-lower-boundᶜ : Conₘ n → Sequence (Conₘ n) → Set a
```
This definition is not documented.

nad commented on Graded/Modality.agda:
```diff
@@ -80,6 +83,19 @@ record Semiring-with-meet : Set a where
   Least-such-that : (M → Set a) → M → Set a
   Least-such-that P p = P p × (∀ q → P q → p ≤ q)
 
+  Greatest-such-that : ∀ {ℓ} → (M → Set ℓ) → M → Set (a ⊔ ℓ)
```
This definition is not documented.

nad commented on Tools/Function.agda:
```diff
@@ -299,3 +301,8 @@ Is-proposition-Dec = λ where
   fe _    {x = no x}  {y = no y}  →
     cong no (Is-proposition-Π fe λ _ → ⊥-propositional)
+
+-- Infinite sequences of elements of a given type
```
I think it would make more sense to have this definition in `Tools.Nat`.

nad commented on Graded/Modality/Instances/Nat-plus-infinity.agda:
```diff
@@ -996,6 +1032,46 @@ opaque
 
+opaque
+
+  nr₃-𝟘 : ∀ r → nr₃ r ⌞ 0 ⌟ ⌞ 0 ⌟ ≡ ⌞ 0 ⌟
```
This definition is not documented.

nad commented on Graded/Modality/Instances/Nat-plus-infinity.agda:
```diff
@@ -996,6 +1032,46 @@ opaque
+
+opaque
+
+  nr₃-+ : ∀ r → nr₃ r z₁ s₁ + nr₃ r z₂ s₂ ≤ nr₃ r (z₁ + z₂) (s₁ + s₂)
```
This definition is not documented.

nad commented on Graded/Modality/Instances/Linear-or-affine.agda:
```diff
@@ -1545,28 +1547,34 @@ linear-or-affine-has-star = record
+  -- these problems (see
+  -- Graded.Modality.Instances.Linear-or-affine.Good), but note that, at
+  -- the time of writing, this formalisation does not contain any solid
+  -- evidence showing that linear-or-affine is "correct".
```
Is that still true?

nad commented on Graded/Usage/Restrictions/Natrec.agda:
```diff
@@ -0,0 +1,165 @@
+open import Tools.PropositionalEquality
+open import Tools.Product
+-- open import Tools.Relation
+-- open import Tools.Sum hiding (sym)
```
The commented-out code above could be removed.

nad commented on Graded/Usage/Restrictions/Natrec.agda:
```diff
@@ -0,0 +1,165 @@
+--     Nr-available-propositional = lemma
+--       where
+--       lemma : (x y : Natrec-mode-has-nr nm) → x ≡ y
+--       lemma Nr Nr = refl
```
More commented-out code.

nad commented on Graded/Usage/Restrictions/Natrec.agda:
```diff
@@ -0,0 +1,165 @@
+--       lemma Nr Nr = refl
+
+
+data Natrec-mode? (nm : Natrec-mode) : Set a where
```
This definition is not documented.

nad replied:
What is the purpose of this definition? Could it be removed?

oskeri replied:
It has essentially the same purpose as `Dedicated-nr?` has/had. It allows case distinction on a given `Natrec-mode` in a way that brings corresponding instance arguments into scope. It is used, for instance, in `Graded.Usage.Properties.natrec-nr-or-no-nrₘ` and a few other places.

It would be possible to split on the `Natrec-mode` directly instead but then one would have to deal with the instance arguments manually.

nad commented on Graded/Usage/Restrictions/Natrec.agda:
```diff
@@ -0,0 +1,165 @@
+
+opaque
+
+  natrec-mode? : ∀ nm → Natrec-mode? nm
```
This definition is not documented.

nad commented on Graded/Usage/Restrictions/Instance.agda:
```diff
@@ -0,0 +1,33 @@
+------------------------------------------------------------------------
+-- ???
```
???

nad commented on Graded/Usage/Restrictions.agda:
```diff
@@ -97,6 +103,18 @@ record Usage-restrictions : Set (lsuc a) where
       erased-matches-for-K 𝟙ᵐ ≤ᵉᵐ erased-matches-for-K 𝟘ᵐ[ ok ]
 
+  -- Three mutually exclusive types which corresponding to each of the
+  -- three poossibilities for natrec-mode
```
Spelling.

nad commented on Graded/Usage/Restrictions/Instance.agda:
```diff
@@ -0,0 +1,33 @@
+
+instance
+
+  Nr-not-available₂-Supports-factoring :
```
What does the 2 stand for?

nad commented on Graded/Restrictions.agda:
```diff
@@ -561,12 +570,12 @@ opaque
   -- If grade equality is decidable and the modality comes with a
   -- dedicated nr function, then UD.Assumptions holds for
-  -- no-usage-restrictions b false.
+  -- no-usage-restrictions Nr b false.
```
The code in the comment is underspecified (note that `Nr` takes an instance argument).

nad commented on Graded/Modality/Instances/Affine/Bad.agda:
```diff
@@ -15,28 +14,38 @@ module Graded.Modality.Instances.Affine.Bad
+  UR′ = nr-available-UR zero-one-many-greatest-star-nr UR
+  open Usage-restrictions UR′
+  instance
+    has-nr : Has-nr M.semiring-with-meet
```
Is `has-nr` only used in the implementation of `nr-avail`? In that case it might make sense to inline `has-nr` (and rename `nr-avail`). However, I tried to do this, and for some reason Agda then failed to instantiate some instance arguments below, even though those arguments could be instantiated using `nr-avail`. I think this should be investigated.

Similar comments apply to other modules below as well.

nad commented on Graded/Modality/Morphism/Usage-restrictions/Examples.agda:
```diff
@@ -109,15 +132,31 @@ opaque
+     ⦃ has-nr₂ : Natrec-mode-has-nr _ nm₂ ⦄ →
+     Is-nr-reflecting-morphism 𝕄₁ 𝕄₂
+       ⦃ Natrec-mode-Has-nr _ has-nr₁ ⦄
+       ⦃ Natrec-mode-Has-nr _ has-nr₂ ⦄ tr) →
```
I suggest that you give instance arguments by name.

nad commented on Graded/Modality/Morphism/Usage-restrictions/Examples.agda:
```diff
@@ -427,7 +481,7 @@ opaque
   -- If the functions tr and tr-Σ preserve certain usage restrictions,
   -- then they also do this for certain usage restrictions obtained
-  -- using not-all-erased-matches-JK.
+  -- using not-all-erased-matches-JK given certain assumptions.
```
What assumptions?

nad commented on Graded/Heap/Soundness.agda:
```diff
@@ -4,20 +4,24 @@
+    ⦃ has-nr : Nr-available ⦄ →
+    Is-factoring-nr M (Natrec-mode-Has-nr 𝕄 has-nr))
+  (open Graded.Heap.Bisimilarity UR TR factoring-nr)
   (As : Assumptions)
```
Here you have one assumption that says that if the rule with an nr function is used, then that nr function is factoring, then you have another assumption *hidden inside `Assumptions`* that says that the `Nr-not-available` rule is not used. In the interests of readability I think this telescope should be refactored. Perhaps you could have a positive assumption inside `Assumptions` that states that either `Nr-not-available-GLB` holds, or `Nr-available` holds for a factoring nr function.

The same comment applies to `Graded.Heap.Termination` as well.

nad commented on Graded/Usage/Decidable/Assumptions.agda:
```diff
@@ -49,7 +48,7 @@ record Assumptions : Set a where
     instance
       -- A dedicated nr function is assumed to exist.
-      ⦃ has-nr ⦄ : Dedicated-nr
+      ⦃ has-nr ⦄ : Nr-available
```
Could decidability also be proved under something like the following assumption?
```agda
Nr-not-available-GLB ×
(∀ r p q → ∃ λ s → Greatest-lower-bound s (nrᵢ r p q))
```
And does the second part of this assumption hold for at least some of the example modalities? (I think you and I have already discussed something like this.)

nad reviewed (commented):
Would it make sense to add modules like `Graded.Modality.Instances.Affine.Good` that use `No-nr-glb` instead of `Nr`?

I would like to know more about how the system with `Nr-available` compares to the one with `Nr-not-available-GLB` (and I know that you have already thought about this).

I don't like having nr functions in the usage restrictions: other modality functions are part of the modality, not a later addition. However, if my question about decidability can be answered sufficiently affirmatively, then one could perhaps see the system with `Nr-not-available-GLB` as the "correct" one, and the system with `Nr-available` as an experimental extension.

oskeri commented on Graded/Usage.agda:
```diff
@@ -83,7 +86,7 @@ infix 50 ⌈_⌉
 mutual
   ⌈_⌉ :
-    ⦃ has-nr : Has-nr semiring-with-meet ⦄ →
+    ⦃ has-nr : Nr-available ⦄ →
```
The reason I put `Nr-available` instead of `Has-nr` is that even though you could define this function given an nr function, it wouldn't say anything meaningful unless the usage rule for natrec using that nr function is used.

As you commented below, it should be possible to have inference also without nr functions (given some assumptions about GLB:s) so I will be making some changes here anyway.

oskeri commented on Graded/Usage.agda:
```diff
@@ -353,6 +354,41 @@ data _▸[_]_ {n : Nat} : (γ : Conₘ n) → Mode → Term n → Set a where
+           → θ ∙ ⌜ 𝟘ᵐ? ⌝ · q ▸[ 𝟘ᵐ? ] A
+           → Greatest-lower-bound x (nrᵢ r 𝟙 p)
+           → Greatest-lower-boundᶜ χ (nrᵢᶜ r γ δ)
+           → x ·ᶜ η +ᶜ χ ▸[ m ] natrec p q r A z s n
```
I don't see how to do this. That would be a more permissive rule.

nad replied:
My idea was to combine this rule with `sub`: any lower bound is bounded by a greatest lower bound. I guess that works if a greatest lower bound can always be computed. Did you investigate whether GLBs can be computed for the example modalities?

oskeri commented on Graded/Modality/Morphism/Examples.agda:
```diff
@@ -1244,6744 +2120,6548 @@ zero-one-many⇨erasure {v₂ = v₂@record{}} {𝟙≤𝟘 = 𝟙≤𝟘} refl
+      𝟘 𝟘 𝟙 𝟘 ω → refl
+      𝟘 𝟘 𝟙 𝟙 𝟘 → refl
+      𝟘 𝟘 𝟙 𝟙 𝟙 → refl
+   
```
For most morphisms, I did not attempt to prove whether they are `Is-no-nr-glb-reflecting-morphism`. This as well as figuring out whether `Is-no-nr-glb-reflecting-morphism` is defined "correctly/suitably" is left as future work (the current definition is more or less exactly what seems to be needed).

oskeri commented on Graded/Modality/Morphism/Examples.agda:
```diff
@@ -1244,6744 +2120,6548 @@ zero-one-many⇨erasure {v₂ = v₂@record{}} {𝟙≤𝟘 = 𝟙≤𝟘} refl
+      𝟘 𝟘 𝟙 𝟘 ω → refl
+      𝟘 𝟘 𝟙 𝟙 𝟘 → refl
+      𝟘 𝟘 𝟙 𝟙 𝟙 → refl
+   
```
Since these are not order preserving morphisms I didn't think proving whether they are `Is-no-nr-glb-reflecting-morphism` would be particularly interesting

oskeri commented:
> Would it make sense to add modules like `Graded.Modality.Instances.Affine.Good` that use `No-nr-glb` instead of `Nr`?

I have added `Graded.Modality.Instances.Affine.Good.Greatest-lower-bound` and similar for linearity and linear or affine.

> I would like to know more about how the system with `Nr-available` compares to the one with `Nr-not-available-GLB` (and I know that you have already thought about this).

There is probably more that can be said about this but I have added some results in `Graded.Modality.Properties.Natrec`

>  if my question about decidability can be answered sufficiently affirmatively

Usage inference/checking is indeed decidable under the assumption you suggested.
